### PR TITLE
[Bugfix] Truncate build logs

### PR DESCRIPTION
**Pull request by kloessst**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

Build logs entries which are larger than 255 characters currently crash the result creation as they don't fit into the database column VARCHAR(255). This happens for example when a build fails with static code analysis enabled, as Checkstyle will output quite a large error message.

### Description

- Truncate the logs to 255 characters if they are larger than 255 characters
- Little restructure of code. Preparation of build logs moved into a private method
- Filtering of useless SCA build logs e.g. "[INFO] Downloading", "[INFO] Downloaded"
- Integration tests for various sizes of build logs

### Steps for Testing

1. Create a programming exercise (with and without SCA)
2. Participate in those exercises
3. Submit valid solutions
--> Should compile and show results
4. Submit solution containing errors
--> Should not compile and build logs should be shown

### Test Coverage

* BambooService.java: 77->80%
* BambooBuildResultNotificationDTO.BuildLogDTO 0% -> 100%

### CODE REVIEW

**Comment by JohannesStoehr** [comment]
While testing this fix, I discovered a more severe problem.
After creating new programming exercises (with and without SCA) changing out the test cases doesn't work any more.
Now Artemis doesn't recognize the deletion of the old test cases or the creation of the new test cases.
So after changing the test cases and submitting some code as a student the feedback is "0 of 13 passed" (13 is the number of tests in the example exercise), even though I only have one test in the test repo for the exercise right now.

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
I don't get any compilation errors displayed anymore, only "Result could not be generated" and "The result for your last submission is not available. Please submit your code again."for SCA and without (doing so does not improve the problem).

As far as I know Johannes had the same problem as well.

**Comment by kloessst** [comment]
Thanks for the feedback I'll look into this

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
- **No SCA:**
   image
- **With SCA:**
   image

For some reason compilation errors still get lost.
_(Tested on TS1, in Chris Test Course, Exercises named after PR.)_

**Comment by kloessst** [comment]
The bamboo notification plugin checks the last 1000 log entries, applies an error filter and sends the remaining entries to Artemis. We already had this issue in the past, when Artemis was pulling the logs. The downloading logs are so numerous, that the actual compilation error won't get sent. Bumping to number to ~2000 should solve this issue.

Additionally we will add the static code analysis tools to the build agent docker images, so that there shouldn't be any downloading messages anymore. I also added filtering for those messages in Artemis.

Generally, this PR does what it is supposed to do. Long build logs don't crash the result creation anymore. For compilation errors to show up when static code analysis is enabled, a new version of the notification plugin must be rolled out. I'm optimistic that we can do this till the end of the week.

**Comment by kloessst** [comment]
> 
> 
> While testing this fix, I discovered a more severe problem.
> After creating new programming exercises (with and without SCA) changing out the test cases doesn't work any more.
> Now Artemis doesn't recognize the deletion of the old test cases or the creation of the new test cases.
> So after changing the test cases and submitting some code as a student the feedback is "0 of 13 passed" (13 is the number of tests in the example exercise), even though I only have one test in the test repo for the exercise right now.

I could not reproduce this error. Did you run the solution build plan after changes the test cases? Otherwise Artemis won't know that the test cases have changed.

**Review by anditurdiu — APPROVED** [review]
Code looks good. Tested on TS1, workes as expected for me. I could not reproduce the error @MaisiKoleni referenced

**Review by MaisiKoleni — APPROVED** [review]
I still have the same problem described in my last review but as this is out of scope for this PR and requires changes elsewhere, this PR solves the initial problem it was meant to solve.  Everything apart from that issue works fine.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java [hunk 1]
```diff
@@ -290,6 +290,9 @@ else if (status.get("isActive") && status.get("isBuilding")) {
 
         var buildLogEntries = filterBuildLogs(retrieveLatestBuildLogsFromBamboo(programmingExerciseParticipation.getBuildPlanId()));
 
+        // Truncate the logs so that they fit into the database
+        buildLogEntries.forEach(BuildLogEntry::truncateLogToMaxLength);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooService.java:294]: follow up: make sure this is also invoked for Jenkins

**Review by krusche — APPROVED** [review]

**Pull request merged**